File: 2026POTD/JANUARY/14-01-2026/3454.rb
# @param {Integer[][]} squares
# @return {Float}
def separate_squares(squares)
  xs = []
  events = []

  squares.each do |x, y, l|
    xs << x
    xs << x + l
    events << [y, x, x + l, 1]
    events << [y + l, x, x + l, -1]
  end

  xs.sort!.uniq!
  m = xs.length
  idx = {}
  xs.each_with_index { |v, i| idx[v] = i }

  cnt = Array.new(4 * m, 0)
  seglen = Array.new(4 * m, 0)

  update = lambda do |i, l, r, ql, qr, v|
    next if qr <= l || r <= ql
    if ql <= l && r <= qr
      cnt[i] += v
    else
      mid = (l + r) / 2
      update.call(i * 2, l, mid, ql, qr, v)
      update.call(i * 2 + 1, mid, r, ql, qr, v)
    end
    if cnt[i] > 0
      seglen[i] = xs[r] - xs[l]
    elsif l + 1 == r
      seglen[i] = 0
    else
      seglen[i] = seglen[i * 2] + seglen[i * 2 + 1]
    end
  end

  events.sort_by!(&:first)

  prev = events[0][0]
  total = 0.0
  i = 0

  while i < events.length
    y = events[i][0]
    dy = y - prev
    total += seglen[1] * dy if dy > 0 && seglen[1] > 0

    while i < events.length && events[i][0] == y
      _, x1, x2, t = events[i]
      update.call(1, 0, m, idx[x1], idx[x2], t)
      i += 1
    end
    prev = y
  end

  half = total / 2.0
  cnt.fill(0)
  seglen.fill(0)

  prev = events[0][0]
  cur = 0.0
  i = 0

  while i < events.length
    y = events[i][0]
    dy = y - prev
    if dy > 0 && seglen[1] > 0
      area = seglen[1] * dy
      return prev + (half - cur) / seglen[1] if cur + area >= half
      cur += area
    end

    while i < events.length && events[i][0] == y
      _, x1, x2, t = events[i]
      update.call(1, 0, m, idx[x1], idx[x2], t)
      i += 1
    end
    prev = y
  end

  prev.to_f
end

File: 2026POTD/JANUARY/14-01-2026/3454.rs
impl Solution {
    pub fn separate_squares(squares: Vec<Vec<i32>>) -> f64 {
        let n = squares.len();

        let mut xs: Vec<i64> = Vec::with_capacity(2 * n);
        #[derive(Clone)]
        struct Event {
            y: i64,
            x1: i64,
            x2: i64,
            t: i64,
        }
        let mut events: Vec<Event> = Vec::with_capacity(2 * n);

        for s in squares {
            let x = s[0] as i64;
            let y = s[1] as i64;
            let l = s[2] as i64;
            xs.push(x);
            xs.push(x + l);
            events.push(Event { y, x1: x, x2: x + l, t: 1 });
            events.push(Event { y: y + l, x1: x, x2: x + l, t: -1 });
        }

        xs.sort_unstable();
        xs.dedup();
        let m = xs.len();

        use std::collections::HashMap;
        let mut idx = HashMap::with_capacity(m);
        for (i, &v) in xs.iter().enumerate() {
            idx.insert(v, i);
        }

        let mut cnt = vec![0i64; 4 * m];
        let mut seglen = vec![0i64; 4 * m];

        fn upd(
            i: usize,
            l: usize,
            r: usize,
            ql: usize,
            qr: usize,
            v: i64,
            xs: &Vec<i64>,
            cnt: &mut Vec<i64>,
            seglen: &mut Vec<i64>,
        ) {
            if qr <= l || r <= ql {
                return;
            }
            if ql <= l && r <= qr {
                cnt[i] += v;
            } else {
                let mid = (l + r) >> 1;
                upd(i << 1, l, mid, ql, qr, v, xs, cnt, seglen);
                upd(i << 1 | 1, mid, r, ql, qr, v, xs, cnt, seglen);
            }
            if cnt[i] > 0 {
                seglen[i] = xs[r] - xs[l];
            } else if l + 1 == r {
                seglen[i] = 0;
            } else {
                seglen[i] = seglen[i << 1] + seglen[i << 1 | 1];
            }
        }

        events.sort_by_key(|e| e.y);

        let mut prev = events[0].y;
        let mut total = 0f64;
        let mut i = 0usize;

        while i < events.len() {
            let y = events[i].y;
            let dy = y - prev;
            if dy > 0 && seglen[1] > 0 {
                total += seglen[1] as f64 * dy as f64;
            }
            while i < events.len() && events[i].y == y {
                let e = &events[i];
                upd(
                    1,
                    0,
                    m,
                    idx[&e.x1],
                    idx[&e.x2],
                    e.t,
                    &xs,
                    &mut cnt,
                    &mut seglen,
                );
                i += 1;
            }
            prev = y;
        }

        let half = total / 2.0;
        cnt.fill(0);
        seglen.fill(0);

        prev = events[0].y;
        let mut cur = 0f64;
        i = 0;

        while i < events.len() {
            let y = events[i].y;
            let dy = y - prev;
            if dy > 0 && seglen[1] > 0 {
                let area = seglen[1] as f64 * dy as f64;
                if cur + area >= half {
                    return prev as f64 + (half - cur) / seglen[1] as f64;
                }
                cur += area;
            }
            while i < events.len() && events[i].y == y {
                let e = &events[i];
                upd(
                    1,
                    0,
                    m,
                    idx[&e.x1],
                    idx[&e.x2],
                    e.t,
                    &xs,
                    &mut cnt,
                    &mut seglen,
                );
                i += 1;
            }
            prev = y;
        }

        prev as f64
    }
}